### ISSUE TITLE
Add a dedicated `SparseGenericTransposeSolve` autograd function for general higher-order sparse solves

### ISSUE THREAD

**Issue by theo-barfoot** [cai4cai/torchsparsegradutils#84]
## Background

Issue #81 exposed a higher-order-gradient failure in `sparse_generic_solve` when using solvers such as `minres`. The immediate bug was that `SparseGenericSolve.backward` called the raw transpose solver directly. With `create_graph=True`, PyTorch attempted to differentiate through the raw iterative solver implementation and failed on unsupported `out=` operations.

A pragmatic fix is to replace the direct raw solver call in `SparseGenericSolve.backward` with a nested `sparse_generic_solve` call:

```python
gradB = sparse_generic_solve(
    A,
    grad,
    solve=ctx.transpose_solve,
    transpose_solve=ctx.solve,
    **ctx.kwargs,
)
```

That is sufficient for the reported issue involving symmetry-oriented solvers such as `minres` and `linear_cg`, and it avoids differentiating through iterative-solver internals.

However, a more principled long-term design would be to introduce a dedicated autograd function for transpose solves.

## Motivation

`SparseGenericSolve` represents the operation:

```text
A X = B
X = A^{-1} B
```

Its backward pass uses:

```text
gradB = A^{-T} grad
```

and computes sparse matrix gradients only at the nonzero entries of `A`.

For higher-order differentiation, the transpose solve appearing in the backward pass should itself be represented as an implicit solve with a custom backward rule. Conceptually, this operation is:

```text
A^T Y = G
Y = A^{-T} G
```

This is not the same operation as `A Y = G` for non-symmetric matrices. For symmetric matrices, the distinction collapses, but for general non-symmetric systems it matters.

## Limitation of the current pragmatic fix

The current nested `sparse_generic_solve` approach is likely appropriate for symmetric solvers such as `minres` and `linear_cg`, where `A^{-T}` and `A^{-1}` coincide for symmetric `A`.

For arbitrary non-symmetric systems with distinct `solve` and `transpose_solve` callables, the transpose-solve operation has a different sparse `gradA` formula from the ordinary solve operation. The row/column indexing in the sparse matrix gradient is flipped.

For a standard solve:

```text
A X = B
```

with upstream gradient `G`, the implicit backward rule involves:

```text
U = A^{-T} G
gradA[i, j] = -U[i, :] · X[j, :]
```

For a transpose solve:

```text
A^T Y = G
```

with upstream gradient `H`, the backward rule should involve:

```text
V = A^{-1} H
gradA[i, j] = -V[j, :] · Y[i, :]
```

That index swap is the main reason a separate `SparseGenericTransposeSolve` would be cleaner and safer than reusing `SparseGenericSolve` for all cases.

## Proposed design

Introduce a new internal custom autograd function, for example:

```python
class SparseGenericTransposeSolve(torch.autograd.Function):
    ...
```

and a small internal wrapper, for example:

```python
def sparse_generic_transpose_solve(
    A: torch.Tensor,
    G: torch.Tensor,
    solve: Callable[..., torch.Tensor],
    transpose_solve: Callable[..., torch.Tensor],
    **kwargs,
) -> torch.Tensor:
    ...
```

This function should represent:

```text
Y = A^{-T} G
```

### Forward pass

The forward pass should call the provided transpose solver:

```python
Y = transpose_solve(A.detach(), G.detach(), **kwargs)
```

and save the tensors needed for backward.

For higher-order support, avoid detaching saved outputs in a way that would break second derivatives.

### Backward pass

Given upstream gradient `H = dL/dY`, compute:

```text
gradG = A^{-1} H
```

using the original forward solver:

```python
gradG = sparse_generic_solve(
    A,
    H,
    solve=solve,
    transpose_solve=transpose_solve,
    **kwargs,
)
```

Then compute the sparse gradient with respect to `A` only at the nonzero entries.

For each nonzero `A[i, j]`, use the transpose-solve gradient pattern:

```text
gradA[i, j] = -gradG[j, :] · Y[i, :]
```

This differs from the ordinary solve case, where the row/column selection is:

```text
gradA[i, j] = -gradB[i, :] · X[j, :]
```

## Integration point

In `SparseGenericSolve.backward`, replace the nested `sparse_generic_solve` transpose solve with the new internal transpose-solve wrapper:

```python
gradB = sparse_generic_transpose_solve(
    A,
    grad,
    solve=ctx.solve,
    transpose_solve=ctx.transpose_solve,
    **ctx.kwargs,
)
```

This would make the code more explicit: `SparseGenericSolve` handles `A X = B`, while `SparseGenericTransposeSolve` handles `A^T Y = G`.

## Acceptance criteria

- Existing `sparse_generic_solve` first-order tests still pass.
- Higher-order tests for `minres` and `linear_cg` still pass.
- New tests cover genuinely non-symmetric systems using distinct `solve` and `transpose_solve` callables.
- COO and CSR layouts are both tested.
- Vector RHS and matrix RHS are both tested.
- The sparse gradient with respect to `A` matches dense `torch.linalg.solve` references at the nonzero entries.
- First derivatives and second-derivative-vector-style checks match dense references within reasonable tolerances.

### 1. Non-symmetric differentiable sparse matrix

Create a small fixed-sparsity non-symmetric matrix whose values depend on a differentiable parameter vector `theta`.

Example structure:

```text
A(theta) = diagonally dominant sparse non-symmetric matrix
```

Use diagonal dominance to avoid singular systems.

### 2. Distinct dense reference

Build both:

```python
A_sparse(theta_sparse)
A_dense(theta_dense)
```

Then compare:

```python
out_sparse = sparse_generic_solve(A_sparse, B, solve=solve, transpose_solve=transpose_solve)
out_dense = torch.linalg.solve(A_dense, B)
```

Compare:

```python
grad_sparse = torch.autograd.grad(loss_sparse, theta_sparse, create_graph=True)[0]
grad_dense = torch.autograd.grad(loss_dense, theta_dense, create_graph=True)[0]
```

and:

```python
hess_vec_sparse = torch.autograd.grad(grad_sparse.sum(), theta_sparse)[0]
hess_vec_dense = torch.autograd.grad(grad_dense.sum(), theta_dense)[0]
```

### 3. Explicit transpose-solve test

Directly test the new internal transpose solve against:

```python
torch.linalg.solve(A_dense.T, G)
```

and compare both first- and second-order derivatives.

## Potential implementation risks

- Accidental recursion between `SparseGenericSolve` and `SparseGenericTransposeSolve`.
- Incorrect row/column indexing for sparse `gradA` in the transpose solve.
- Breaking vector RHS shape restoration.
- Subtle detach/save-for-backward mistakes that allow first-order tests to pass but break second-order correctness.
- Tolerance sensitivity for iterative solvers in second-order tests.

## Suggested implementation strategy

1. Keep the existing pragmatic fix for #81 in place.
2. Add `SparseGenericTransposeSolve` as an internal helper rather than changing public API immediately.
3. Test the helper directly against dense references on small CPU `float64` systems.
4. Once correct, route `SparseGenericSolve.backward` through the helper.
5. Only then consider exposing a public `sparse_generic_transpose_solve` API, if useful.

## Why this should be separate from the immediate #81 fix

The #81 fix is a narrow regression fix for users currently blocked by `create_graph=True` with `minres` / `linear_cg`.

A dedicated `SparseGenericTransposeSolve` is a broader design improvement for full generality, especially for non-symmetric systems. It deserves separate review, targeted non-symmetric tests, and careful validation against dense references.

Keeping these separate gives users the quick fix now without hiding a more ambitious autograd redesign inside a small bug-fix PR.

**Comment by tvercaut** [comment]
I think this may be an overkill. The suggestion in https://github.com/cai4cai/torchsparsegradutils/issues/81#issuecomment-4712945726 should already handle non-symmetric cases. Indeed, note that I suggested passing `solve=ctx.transpose_solve` as argument to `sparse_generic_solve` so this already provides a transposed solve.

If we want a nicer more intuitive syntax in `SparseGenericSolve.backward` we can probably just define a thin wrapper:
```python
def sparse_generic_transpose_solve(
    A: torch.Tensor,
    B: torch.Tensor,
    solve: Optional[Callable[..., torch.Tensor]] = None,
    transpose_solve: Optional[Callable[..., torch.Tensor]] = None,
    **kwargs,
) -> torch.Tensor:
    return sparse_generic_solve(A, B, solve=transpose_solve, transpose_solve=solve, **kwargs)
```
